content/posts/2023-09-19-Diagnostic-Improvements-in-Clang-17.md

```diff
@@ -0,0 +1,560 @@
+Moreover, line numbers are now attached to the left of each line, allowing for quicker navigation and issue resolution.
+
+```c++
+int func(
+  int a, int b, int& r);
+
+void test(int *ptr) {
+  func(3, 4, 5);
+  func(3, 4);
+}
+```
+
+Before:
+```console
+<source>:5:3: error: no matching function for call to 'func'
+  func(3, 4, 5);
+  ^~~~
+<source>:1:5: note: candidate function not viable: expects an lvalue for 3rd argument
+int func(
+    ^
+<source>:6:3: error: no matching function for call to 'func'
+  func(3, 4);
+  ^~~~
+<source>:1:5: note: candidate function not viable: requires 3 arguments, but 2 were provided
+int func(
+    ^
+```
+
+After:
+```console
+<source>:5:3: error: no matching function for call to 'func'
+    5 |   func(3, 4, 5);
+      |   ^~~~
+<source>:1:5: note: candidate function not viable: expects an lvalue for 3rd argument
+    1 | int func(
+      |     ^
+    2 |   int a, int b, int& r);
+      |                 ~~~~~~
+<source>:6:3: error: no matching function for call to 'func'
+    6 |   func(3, 4);
+      |   ^~~~
+<source>:1:5: note: candidate function not viable: requires 3 arguments, but 2 were provided
+    1 | int func(
+      |     ^
+    2 |   int a, int b, int& r);
+      |   ~~~~~~~~~~~~~~~~~~~~
+```
+
+In this example, the newly covered source ranges make it easier to understand why the overload candidate is invalid.
+
+Commit: https://reviews.llvm.org/D147875 (Timm Bäder)
+
+### Preprocessor-related diagnostics
+- Clang warns on macro redefinitions. When the redefinition happens in assembly files, and the previous definition of the macro comes from the command line, the last definition is now diagnosed as coming from `<command line>` instead of `<built-in>`.
+
+Assembly file:
+```asm
+#define MACRO 3
+```
+
+Clang invocation command:
+```console
+clang -DMACRO=1 file.S
+```
+Before:
+```console
+warning: 'MACRO' macro redefined [-Wmacro-redefined]
+#define MACRO 3
+        ^
+<built-in>:362:9: note: previous definition is here
+#define MACRO 1
+        ^
+```
+After:
+```console
+warning: 'MACRO' macro redefined [-Wmacro-redefined]
+    1 | #define MACRO 3
+      |         ^
+<command line>:1:9: note: previous definition is here
+    1 | #define MACRO 1
+      |         ^
+```
+Commit: https://reviews.llvm.org/D145397 (John Brawn)
+
+<br>
+
+- Clang 17 emits a warning on any compiler-builtin macro being undefined or redefined, some of which were just ignored in Clang 16.
+
+```c++
+#undef __cplusplus
+```
+
+Before: _No Warning_
+
+After:
+```console
+<source>:1:8: warning: undefining builtin macro [-Wbuiltin-macro-redefined]
+    1 | #undef __cplusplus
+      |        ^
+```
+
+Redefinition of compiler builtin macros usually leads to unintended results because library headers often rely on these macros, and they do not
+expect these macros to be modified by users.
+
+Commit: https://reviews.llvm.org/D144654 (John Brawn)
```

Comment:
I think a warning is a great first step, but these ones should eventually graduate to errors in 2-3 versions. May be worth teasing that here so folks know it's a possibility?

Reply:
Yikes! Gonna move this to Discord so it doesn't clog up the review.